src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/LockFileUtils.cs

```diff
@@ -558,80 +559,85 @@ private static void AddDependencies(IEnumerable<LibraryDependency> dependencies,
-
                     // Find the project path, this is provided by the resolver
                     var msbuildFilePathInfo = new FileInfo((string)msbuildPath);
 
                     // Ensure a trailing slash for the relative path helper.
                     var projectDir = PathUtility.EnsureTrailingSlash(msbuildFilePathInfo.Directory.FullName);
 
-                    // Read files from the project if they were provided.
-                    if (localMatch.LocalLibrary.Items.TryGetValue(KnownLibraryProperties.ProjectRestoreMetadataFiles, out filesObject))
-                    {
-                        files.AddRange((List<ProjectRestoreMetadataFile>)filesObject);
-                    }
-
-                    var targetFrameworkShortName = targetGraph.Framework.GetShortFolderName();
-                    var libAnyPath = $"lib/{targetFrameworkShortName}/any.dll";
-
-                    if (files.Count == 0)
-                    {
-                        // If the project did not provide a list of assets, add in default ones.
-                        // These are used to detect transitive vs non-transitive project references.
-                        var absolutePath = Path.Combine(projectDir, "bin", "placeholder", $"{localMatch.Library.Name}.dll");
+                    // Create an ordered list of selection criteria. Each will be applied, if the result is empty
+                    // fallback frameworks from "imports" will be tried.
+                    // These are only used for framework/RID combinations where content model handles everything.
+                    var orderedCriteria = CreateCriteria(targetGraph.Conventions, targetGraph.Framework, targetGraph.RuntimeIdentifier);
 
-                        files.Add(new ProjectRestoreMetadataFile(libAnyPath, absolutePath));
-                    }
+                    string libAnyPath = $"lib/{targetGraph.Framework.GetShortFolderName()}/any.dll";
+                    var contentItems = new ContentItemCollection();
 
-                    // Process and de-dupe files
-                    for (var i = 0; i < files.Count; i++)
+                    if (localMatch.LocalLibrary.Items.TryGetValue(KnownLibraryProperties.ProjectRestoreMetadataFiles, out object filesObject))
                     {
-                        var path = files[i].PackagePath;
-
-                        // LIBANY avoid compatibility checks and will always be used.
-                        if (LIBANY.Equals(path, StringComparison.Ordinal))
+                        List<ProjectRestoreMetadataFile> files = (List<ProjectRestoreMetadataFile>)filesObject;
+                        if (files.Count > 0)
                         {
-                            path = libAnyPath;
-                        }
+                            var fileLookup = new Dictionary<string, ProjectRestoreMetadataFile>(StringComparer.OrdinalIgnoreCase);
```

Comment:
I'm surprised there could be duplicates in this list?  But that's what the old code was doing, do you know why?